include/mbgl/style/expression/expression.hpp

```diff
@@ -0,0 +1,223 @@
+template <class V>
+ParseResult parseExpression(const V& value, const ParsingContext& context);
+
+using TypecheckResult = variant<std::vector<CompileError>, std::unique_ptr<Expression>>;
+
+using namespace mbgl::style::conversion;
+
+class LiteralExpression : public Expression {
+public:
+    LiteralExpression(std::string key_, Value value_) : Expression(key_, typeOf(value_)), value(value_) {}
+    
+    Value getValue() const { return value; }
+
+    EvaluationResult evaluate(const EvaluationParameters&) const override {
+        return value;
+    }
+    
+    template <class V>
+    static ParseResult parse(const V& value, const ParsingContext& ctx) {
+        const Value& parsedValue = parseValue(value);
+        return std::make_unique<LiteralExpression>(ctx.key(), parsedValue);
+    }
+    
+private:
+    template <class V>
+    static Value parseValue(const V& value) {
+        if (isUndefined(value)) return Null;
+        if (isObject(value)) {
+            std::unordered_map<std::string, Value> result;
+            eachMember(value, [&] (const std::string& k, const V& v) -> optional<conversion::Error> {
+                result.emplace(k, parseValue(v));
+                return {};
+            });
+            return result;
+        }
+        if (isArray(value)) {
+            std::vector<Value> result;
+            const auto length = arrayLength(value);
+            for(std::size_t i = 0; i < length; i++) {
+                result.emplace_back(parseValue(arrayMember(value, i)));
+            }
+            return result;
+        }
+        
+        optional<mbgl::Value> v = toValue(value);
+        assert(v);
+        return convertValue(*v);
+    }
+
+    Value value;
+};
+
+struct NArgs {
+    std::vector<type::Type> types;
+    optional<std::size_t> N;
+};
+
+class LambdaExpression : public Expression {
+public:
+    using Params = std::vector<variant<type::Type, NArgs>>;
+    using Args = std::vector<std::unique_ptr<Expression>>;
+
+    LambdaExpression(std::string key_,
+                    std::string name_,
+                    Args args_,
+                    type::Type type_,
+                    std::vector<Params> signatures_) :
+        Expression(key_, type_),
+        args(std::move(args_)),
+        signatures(signatures_),
+        name(name_)
+    {}
+    
+    std::string getName() const { return name; }
+    
+    virtual bool isFeatureConstant() const override {
+        bool isFC = true;
+        for (const auto& arg : args) {
+            isFC = isFC && arg->isFeatureConstant();
+        }
+        return isFC;
+    }
+    
+    virtual bool isZoomConstant() const override {
+        bool isZC = true;
+        for (const auto& arg : args) {
+            isZC = isZC && arg->isZoomConstant();
+        }
+        return isZC;
+    }
+    
+    virtual std::unique_ptr<Expression> applyInferredType(const type::Type& type, Args args) const = 0;
```

Comment:
This seems like a bad design: `e.applyInferredType` wants to return a transformed copy of `e`, but because it transfers ownership of child expressions, it's destructive of `e` in the process.  Either expressions should be made copyable or `e.applyInferredType` should just mutate `e` in place.